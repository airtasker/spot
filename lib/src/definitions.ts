import { Type, TypeDef } from "./types";

export interface Contract {
  name: string;
  description?: string;
  version?: string;
  config: Config;
  types: { name: string; typeDef: TypeDef }[];
  security?: SecurityHeader;
  endpoints: Endpoint[];
  oa3servers?: Oa3Server[];
}

export interface Config {
  paramSerializationStrategy: {
    query: {
      array: QueryParamArrayStrategy;
    };
  };
}

export interface SecurityHeader {
  name: string;
  description?: string;
  type: Type;
}

export interface Endpoint {
  name: string;
  description?: string;
  summary?: string;
  tags: string[];
  method: HttpMethod;
  path: string;
  request?: Request;
  responses: Response[];
  defaultResponse?: DefaultResponse;
  draft: boolean;
}

export interface Request {
  headers: Header[];
  pathParams: PathParam[];
  queryParams: QueryParam[];
  body?: Body;
}

export interface Response {
  status: number;
  description?: string;
  headers: Header[];
  body?: Body;
}

export type DefaultResponse = Omit<Response, "status">;

export interface Header {
  name: string;
  description?: string;
  type: Type;
  optional: boolean;
  examples?: Example[];
}

export interface PathParam {
  name: string;
  description?: string;
  type: Type;
  examples?: Example[];
}

export interface Example {
  name: string;
  value: any; // TODO: encapsulate type information
}

export interface QueryParam {
  name: string;
  description?: string;
  type: Type;
  optional: boolean;
  examples?: Example[];
}

export interface Body {
  type: Type;
}

export interface Oa3Server {
  url: string;
  description?: string;
  oa3ServerVariables: Oa3ServerVariable[];
}

export interface Oa3ServerVariable {
  type: Type;
  description?: string;
  defaultValue: string;
  parameterName: string;
}

/**
 * Supported serialization strategies for arrays in query parameters
 *
 *    "ampersand": ?id=3&id=4&id=5
 *    "comma": ?id=3,4,5
 */
export type QueryParamArrayStrategy = "ampersand" | "comma";

/** Supported HTTP methods */
export type HttpMethod = "GET" | "POST" | "PUT" | "PATCH" | "DELETE" | "HEAD";

/** Type guards */
export function isSpecificResponse(
  response: DefaultResponse
): response is Response {
  return "status" in response;
}
