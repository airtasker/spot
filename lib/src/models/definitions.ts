import { HttpMethod } from "./http";
import { DataExpression, DataType } from "./types";

export interface ContractDefinition {
  api: ApiDefinition;
  endpoints: EndpointDefinition[];
  types: TypeDefinition[];
}

export interface TypeDefinition {
  name: string;
  description?: string;
  type: DataType;
}

export interface ApiDefinition {
  name: string;
  description?: string;
  securityHeader?: SecurityHeaderDefinition;
}

export interface SecurityHeaderDefinition {
  name: string;
  description?: string;
  type: DataType;
}

export interface EndpointDefinition {
  name: string;
  description?: string;
  tags: string[];
  method: HttpMethod;
  path: string;
  request: RequestDefinition;
  responses: ResponseDefinition[];
  defaultResponse?: DefaultResponseDefinition;
  tests: TestDefinition[];
}

export interface RequestDefinition {
  headers: HeaderDefinition[];
  pathParams: PathParamDefinition[];
  queryParams: QueryParamDefinition[];
  body?: BodyDefinition;
}

/** A response inherits all the properties of default response, as well as specifying a specific status code. */
export interface ResponseDefinition extends DefaultResponseDefinition {
  status: number;
}

/** The default response, is the assumed response when no status code is specified. */
export interface DefaultResponseDefinition {
  description?: string;
  headers: HeaderDefinition[];
  body?: BodyDefinition;
}

export interface HeaderDefinition {
  name: string;
  description?: string;
  type: DataType;
  optional: boolean;
}

export interface PathParamDefinition {
  name: string;
  description?: string;
  type: DataType;
}

export interface QueryParamDefinition {
  name: string;
  description?: string;
  type: DataType;
  optional: boolean;
}

export interface BodyDefinition {
  description?: string;
  type: DataType;
}

export interface TestDefinition {
  description?: string;
  states: TestStateDefinition[];
  request?: TestRequestDefinition;
  response: TestResponseDefinition;
}

export interface TestStateDefinition {
  name: string;
  params: { name: string; expression: DataExpression }[];
}

export interface TestRequestDefinition {
  headers: { name: string; expression: DataExpression }[];
  pathParams: {
    name: string;
    expression: DataExpression;
  }[];
  queryParams: {
    name: string;
    expression: DataExpression;
  }[];
  body?: DataExpression;
}

export interface TestResponseDefinition {
  status: number;
  headers: { name: string; expression: DataExpression }[];
  body?: DataExpression;
}
