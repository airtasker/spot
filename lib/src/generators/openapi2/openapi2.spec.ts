import { isOpenApiv2, Spectral } from "@stoplight/spectral";
import { Contract } from "../../definitions";
import { parseContract } from "../../parsers/contract-parser";
import { createProjectFromExistingSourceFile } from "../../spec-helpers/helper";
import { generateOpenAPI2 } from "./openapi2";

describe("OpenAPI 2 generator", () => {
  const spectral = new Spectral();

  beforeAll(async () => {
    spectral.registerFormat("oas2", isOpenApiv2);
    await spectral.loadRuleset(`${__dirname}/spectral.ruleset.yml`);
  });

  test("minimal contract produces minimal OpenAPI 2", async () => {
    const contract = generateContract("minimal-contract.ts");
    const result = generateOpenAPI2(contract);

    expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
    const spectralResult = await spectral.run(result);
    expect(spectralResult).toHaveLength(0);
  });

  describe("security", () => {
    test("contract with security header", async () => {
      const contract = generateContract("contract-with-security-header.ts");
      const result = generateOpenAPI2(contract);

      expect(result.consumes).toEqual(["application/json"]);
      expect(result.produces).toEqual(["application/json"]);
      expect(result.security).toHaveLength(1);
      expect(result.security).toContainEqual({ SecurityHeader: [] });
      expect(result.securityDefinitions).toHaveProperty("SecurityHeader", {
        type: "apiKey",
        in: "header",
        name: "security-header"
      });
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });
  });

  describe("HTTP verbs", () => {
    test("GET endpoint", async () => {
      const contract = generateContract("contract-with-get-endpoint.ts");
      const result = generateOpenAPI2(contract);

      expect(result.paths["/users"]).toMatchObject({
        get: expect.anything()
      });
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });

    test("POST endpoint", async () => {
      const contract = generateContract("contract-with-post-endpoint.ts");
      const result = generateOpenAPI2(contract);

      expect(result.paths["/users"]).toMatchObject({
        post: expect.anything()
      });
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });

    test("PUT endpoint", async () => {
      const contract = generateContract("contract-with-put-endpoint.ts");
      const result = generateOpenAPI2(contract);

      expect(result.paths["/users/{id}"]).toMatchObject({
        put: expect.anything()
      });
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });

    test("PATCH endpoint", async () => {
      const contract = generateContract("contract-with-patch-endpoint.ts");
      const result = generateOpenAPI2(contract);

      expect(result.paths["/users/{id}"]).toMatchObject({
        patch: expect.anything()
      });
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });

    test("DELETE endpoint", async () => {
      const contract = generateContract("contract-with-delete-endpoint.ts");
      const result = generateOpenAPI2(contract);

      expect(result.paths["/users/{id}"]).toMatchObject({
        delete: expect.anything()
      });
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });
  });

  describe("path params", () => {
    test("endpoint with path params", async () => {
      const contract = generateContract("contract-with-path-params.ts");
      const result = generateOpenAPI2(contract);

      expect(
        result.paths["/companies/{companyId}/users/{userId}"].get
      ).toHaveProperty("parameters", [
        {
          name: "companyId",
          in: "path",
          required: true,
          type: expect.anything()
        },
        {
          name: "userId",
          in: "path",
          required: true,
          type: expect.anything()
        }
      ]);
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });
  });

  describe("query params", () => {
    test("endpoint with query params", async () => {
      const contract = generateContract("contract-with-query-params.ts");
      const result = generateOpenAPI2(contract);

      expect(result.paths["/companies"].get).toHaveProperty("parameters", [
        {
          name: "country",
          in: "query",
          required: true,
          type: expect.anything()
        },
        {
          name: "postcode",
          in: "query",
          required: false,
          type: expect.anything()
        }
      ]);
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });

    test("endpoint with array query param", async () => {
      const contract = generateContract("contract-with-array-query-param.ts");
      const result = generateOpenAPI2(contract);

      expect(result.paths["/companies"].get).toHaveProperty("parameters", [
        {
          name: "countries",
          in: "query",
          required: true,
          type: "array",
          items: {
            type: "string"
          },
          collectionFormat: "multi"
        }
      ]);
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });

    test("endpoint with array query param and comma serialization strategy", async () => {
      const contract = generateContract(
        "contract-with-array-query-param-and-comma-serialization-strategy.ts"
      );
      const result = generateOpenAPI2(contract);

      expect(result.paths["/companies"].get).toHaveProperty("parameters", [
        {
          name: "countries",
          in: "query",
          required: true,
          type: "array",
          items: {
            type: "string"
          },
          collectionFormat: "csv"
        }
      ]);
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });

    test("endpoint with object query param", async () => {
      const contract = generateContract("contract-with-object-query-param.ts");
      expect(() => generateOpenAPI2(contract)).toThrowError("");
    });
  });

  describe("headers", () => {
    test("endpoint with request headers", async () => {
      const contract = generateContract("contract-with-request-headers.ts");
      const result = generateOpenAPI2(contract);

      expect(result.paths["/users"].get).toHaveProperty("parameters", [
        {
          name: "Accept-Encoding",
          in: "header",
          required: false,
          type: expect.anything()
        },
        {
          name: "Accept-Language",
          in: "header",
          required: true,
          type: expect.anything()
        }
      ]);
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });

    test("endpoint with response headers", async () => {
      const contract = generateContract("contract-with-response-headers.ts");
      const result = generateOpenAPI2(contract);

      expect(result.paths["/users"].post).toHaveProperty(
        "responses.201.headers",
        {
          Link: {
            type: expect.anything()
          },
          Location: {
            type: expect.anything()
          }
        }
      );
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });
  });

  describe("responses", () => {
    test("endpoint specific and default responses", async () => {
      const contract = generateContract(
        "contract-with-specific-and-default-responses.ts"
      );
      const result = generateOpenAPI2(contract);

      expect(result.paths["/users/{id}"].get).toHaveProperty("responses", {
        "200": {
          description: expect.anything(),
          schema: expect.anything()
        },
        "404": {
          description: expect.anything(),
          schema: expect.anything()
        },
        default: {
          description: expect.anything(),
          schema: expect.anything()
        }
      });
      expect(JSON.stringify(result, null, 2)).toMatchSnapshot();
      const spectralResult = await spectral.run(result);
      expect(spectralResult).toHaveLength(0);
    });
  });
});

/**
 * Helper function to create contract from a file. Usable
 * only from this directory.
 *
 * @param filename name of the file
 */
function generateContract(filename: string): Contract {
  const file = createProjectFromExistingSourceFile(
    `${__dirname}/__spec-examples__/${filename}`
  ).file;

  const { contract } = parseContract(file).unwrapOrThrow();
  return contract;
}
