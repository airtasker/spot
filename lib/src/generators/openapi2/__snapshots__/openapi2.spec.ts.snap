// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI 2 generator HTTP verbs DELETE endpoint 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users/{id}\\": {
      \\"delete\\": {
        \\"operationId\\": \\"DeleteEndpoint\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"id\\",
            \\"in\\": \\"path\\",
            \\"required\\": true,
            \\"type\\": \\"string\\"
          }
        ],
        \\"responses\\": {
          \\"204\\": {
            \\"description\\": \\"204 response\\"
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator HTTP verbs GET endpoint 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users\\": {
      \\"get\\": {
        \\"operationId\\": \\"GetEndpoint\\",
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"id\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"name\\": {
                    \\"type\\": \\"string\\"
                  }
                },
                \\"required\\": [
                  \\"id\\",
                  \\"name\\"
                ]
              }
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator HTTP verbs PATCH endpoint 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users/{id}\\": {
      \\"patch\\": {
        \\"operationId\\": \\"PatchEndpoint\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"id\\",
            \\"in\\": \\"path\\",
            \\"required\\": true,
            \\"type\\": \\"string\\"
          },
          {
            \\"in\\": \\"body\\",
            \\"name\\": \\"Body\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"name\\"
              ]
            }
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"id\\": {
                  \\"type\\": \\"string\\"
                },
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"id\\",
                \\"name\\"
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator HTTP verbs POST endpoint 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users\\": {
      \\"post\\": {
        \\"operationId\\": \\"PostEndpoint\\",
        \\"parameters\\": [
          {
            \\"in\\": \\"body\\",
            \\"name\\": \\"Body\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"name\\"
              ]
            }
          }
        ],
        \\"responses\\": {
          \\"201\\": {
            \\"description\\": \\"201 response\\",
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"id\\": {
                  \\"type\\": \\"string\\"
                },
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"id\\",
                \\"name\\"
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator HTTP verbs PUT endpoint 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users/{id}\\": {
      \\"put\\": {
        \\"operationId\\": \\"PutEndpoint\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"id\\",
            \\"in\\": \\"path\\",
            \\"required\\": true,
            \\"type\\": \\"string\\"
          },
          {
            \\"in\\": \\"body\\",
            \\"name\\": \\"Body\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"name\\"
              ]
            }
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"id\\": {
                  \\"type\\": \\"string\\"
                },
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"id\\",
                \\"name\\"
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator contract with version produces a versioned OpenAPI 2 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"versioned-contract\\",
    \\"version\\": \\"0.1.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {}
}"
`;

exports[`OpenAPI 2 generator headers endpoint with request headers 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users\\": {
      \\"get\\": {
        \\"operationId\\": \\"EndpointWithRequestHeaders\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"Accept-Encoding\\",
            \\"in\\": \\"header\\",
            \\"required\\": false,
            \\"type\\": \\"string\\"
          },
          {
            \\"name\\": \\"Accept-Language\\",
            \\"in\\": \\"header\\",
            \\"required\\": true,
            \\"type\\": \\"string\\"
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"id\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"name\\": {
                    \\"type\\": \\"string\\"
                  }
                },
                \\"required\\": [
                  \\"id\\",
                  \\"name\\"
                ]
              }
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator headers endpoint with response headers 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users\\": {
      \\"post\\": {
        \\"operationId\\": \\"EndpointWithResponseHeaders\\",
        \\"parameters\\": [
          {
            \\"in\\": \\"body\\",
            \\"name\\": \\"Body\\",
            \\"required\\": true,
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"name\\"
              ]
            }
          }
        ],
        \\"responses\\": {
          \\"201\\": {
            \\"description\\": \\"201 response\\",
            \\"headers\\": {
              \\"Location\\": {
                \\"type\\": \\"string\\"
              },
              \\"Link\\": {
                \\"type\\": \\"string\\"
              }
            },
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"id\\": {
                  \\"type\\": \\"string\\"
                },
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"id\\",
                \\"name\\"
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator intersection types evaluates intersection type 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users\\": {
      \\"get\\": {
        \\"operationId\\": \\"IntersectionType\\",
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"allOf\\": [
                {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"id\\": {
                      \\"type\\": \\"string\\"
                    }
                  },
                  \\"required\\": [
                    \\"id\\"
                  ]
                },
                {
                  \\"type\\": \\"object\\",
                  \\"properties\\": {
                    \\"name\\": {
                      \\"type\\": \\"string\\"
                    }
                  },
                  \\"required\\": [
                    \\"name\\"
                  ]
                }
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator minimal contract produces minimal OpenAPI 2 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {}
}"
`;

exports[`OpenAPI 2 generator path params endpoint with path params 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/companies/{companyId}/users/{userId}\\": {
      \\"get\\": {
        \\"operationId\\": \\"EndpointWithPathParams\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"companyId\\",
            \\"in\\": \\"path\\",
            \\"required\\": true,
            \\"type\\": \\"string\\"
          },
          {
            \\"name\\": \\"userId\\",
            \\"in\\": \\"path\\",
            \\"required\\": true,
            \\"type\\": \\"string\\"
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"id\\": {
                  \\"type\\": \\"string\\"
                },
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"id\\",
                \\"name\\"
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator query params endpoint with array query param 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/companies\\": {
      \\"get\\": {
        \\"operationId\\": \\"EndpointWithArrayQueryParam\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"countries\\",
            \\"in\\": \\"query\\",
            \\"required\\": true,
            \\"type\\": \\"array\\",
            \\"collectionFormat\\": \\"multi\\",
            \\"items\\": {
              \\"type\\": \\"string\\"
            }
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"id\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"name\\": {
                    \\"type\\": \\"string\\"
                  }
                },
                \\"required\\": [
                  \\"id\\",
                  \\"name\\"
                ]
              }
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator query params endpoint with array query param and comma serialization strategy 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/companies\\": {
      \\"get\\": {
        \\"operationId\\": \\"EndpointWithArrayQueryParam\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"countries\\",
            \\"in\\": \\"query\\",
            \\"required\\": true,
            \\"type\\": \\"array\\",
            \\"collectionFormat\\": \\"csv\\",
            \\"items\\": {
              \\"type\\": \\"string\\"
            }
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"id\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"name\\": {
                    \\"type\\": \\"string\\"
                  }
                },
                \\"required\\": [
                  \\"id\\",
                  \\"name\\"
                ]
              }
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator query params endpoint with query params 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/companies\\": {
      \\"get\\": {
        \\"operationId\\": \\"EndpointWithQueryParams\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"country\\",
            \\"in\\": \\"query\\",
            \\"required\\": true,
            \\"type\\": \\"string\\"
          },
          {
            \\"name\\": \\"post.code\\",
            \\"in\\": \\"query\\",
            \\"required\\": false,
            \\"type\\": \\"string\\"
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"id\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"name\\": {
                    \\"type\\": \\"string\\"
                  }
                },
                \\"required\\": [
                  \\"id\\",
                  \\"name\\"
                ]
              }
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator responses endpoint specific and default responses 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users/{id}\\": {
      \\"get\\": {
        \\"operationId\\": \\"EndpointWithResponses\\",
        \\"parameters\\": [
          {
            \\"name\\": \\"id\\",
            \\"in\\": \\"path\\",
            \\"required\\": true,
            \\"type\\": \\"string\\"
          }
        ],
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"id\\": {
                  \\"type\\": \\"string\\"
                },
                \\"name\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"id\\",
                \\"name\\"
              ]
            }
          },
          \\"404\\": {
            \\"description\\": \\"404 response\\",
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"message\\": {
                  \\"type\\": \\"string\\"
                },
                \\"status\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"message\\",
                \\"status\\"
              ]
            }
          },
          \\"default\\": {
            \\"description\\": \\"default response\\",
            \\"schema\\": {
              \\"type\\": \\"object\\",
              \\"properties\\": {
                \\"message\\": {
                  \\"type\\": \\"string\\"
                }
              },
              \\"required\\": [
                \\"message\\"
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`OpenAPI 2 generator security contract with security header 1`] = `
"{
  \\"swagger\\": \\"2.0\\",
  \\"info\\": {
    \\"title\\": \\"contract\\",
    \\"version\\": \\"0.0.0\\"
  },
  \\"consumes\\": [
    \\"application/json\\"
  ],
  \\"produces\\": [
    \\"application/json\\"
  ],
  \\"paths\\": {
    \\"/users\\": {
      \\"get\\": {
        \\"operationId\\": \\"GetEndpoint\\",
        \\"responses\\": {
          \\"200\\": {
            \\"description\\": \\"200 response\\",
            \\"schema\\": {
              \\"type\\": \\"array\\",
              \\"items\\": {
                \\"type\\": \\"object\\",
                \\"properties\\": {
                  \\"id\\": {
                    \\"type\\": \\"string\\"
                  },
                  \\"name\\": {
                    \\"type\\": \\"string\\"
                  }
                },
                \\"required\\": [
                  \\"id\\",
                  \\"name\\"
                ]
              }
            }
          }
        }
      }
    }
  },
  \\"securityDefinitions\\": {
    \\"SecurityHeader\\": {
      \\"type\\": \\"apiKey\\",
      \\"in\\": \\"header\\",
      \\"name\\": \\"security-header\\"
    }
  },
  \\"security\\": [
    {
      \\"SecurityHeader\\": []
    }
  ]
}"
`;
